services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/utils/MongoDbDeviceRegistryUtils.java

```diff
@@ -180,4 +191,117 @@ public static boolean isDuplicateKeyError(final Throwable error) {
         }
         return false;
     }
+
+    /**
+     * Finds resources such as tenant or device from the given MongoDB collection with the provided 
+     * paging, filtering and sorting options.
+     * <p>
+     * A MongoDB aggregation pipeline is used to find the resources from the given MongoDB collection.
+     *
+     * @param mongoClient The client for accessing the Mongo DB instance.
+     * @param collectionName The name of the collection.
+     * @param pageSize The maximum number of results to include in a response.
+     * @param pageOffset The offset into the result set from which to include objects in the response.
+     *                   This allows to retrieve the whole result set page by page.
+     * @param filterDocument The document used for filtering the resources in a MongoDB aggregation pipeline.
+     * @param sortDocument The document used for sorting the resources in a MongoDB aggregation pipeline.
+     * @param resultMapper The mapper used for mapping the result for the search operation.
+     * @param <T> The type of the result namely {@link org.eclipse.hono.service.management.device.DeviceWithId} or 
+     *           {@link org.eclipse.hono.service.management.tenant.TenantWithId}
+     * @return A future indicating the outcome of the operation. The future will succeed if the search operation 
+     *         is successful and some resources are found. Otherwise the future will fail with the error reason.
```

Comment:
it would be helpful to know what type of exception to expect in a failed future

Reply:
Sure. I have updated it now and pushed a new commit for that.